cdap-ui/app/features/tracker/controllers/tags-ctrl.js

```diff
@@ -20,67 +20,215 @@ class TrackerTagsController{
-        },
-        {
-          'key': 'tag4',
-          'total_entity': 6
-        },
-        {
-          'key': 'tag5',
-          'total_entity': 28
+
+        angular.forEach((response.userTags), (count, tag) => {
+          this.tags.userTags.push({
+            name: tag,
+            count: count,
+            empty: false
+          });
+        });
+        // Calculate number of empty rows needed so that table maintains a consistent height
+        if ((this.tags.userTags.length % 10) !== 0) {
+          let emptyRows = 10 - (this.tags.userTags.length % 10);
+          this.tags.emptyUserTags = {
+            emptyArr: Array(emptyRows),
+            numPages: Math.ceil(this.tags.userTags.length / 10)
+          };
         }
-      ]
+      }, (err) => {
+        console.log('Error', err);
+      });
+  }
+
+  promoteUserTags(tag) {
+    let params = {
+      namespace: this.$state.params.namespace,
+      scope: this.$scope
     };
+
+    this.myTrackerApi.promoteUserTags(params, [tag])
+      .$promise
+      .then((response) => {
+        this.response = response;
+        this.fetchTags();
+      }, (err) => {
+        console.log('Error', err);
+      });
   }
 
-  addPreferredTags() {
-    this.$uibModal.open({
+  demotePreferredTags(tag) {
+    let params = {
+      namespace: this.$state.params.namespace,
+      scope: this.$scope
+    };
+
+    this.myTrackerApi.demotePreferredTags(params, [tag])
+      .$promise
+      .then((response) => {
+        this.response = response;
+        this.fetchTags();
+      }, (err) => {
+        console.log('Error', err);
+      });
+  }
+
+  showDeleteModal(tag) {
+    let modal = this.$uibModal.open({
+      templateUrl: '/assets/features/tracker/templates/partial/delete-preferred-tags-popover.html',
+      size: 'sm',
+      backdrop: true,
+      keyboard: true,
+      windowTopClass: 'tracker-modal delete-modal',
+      controller: DeletePreferredTagsModalCtrl,
+      controllerAs: 'DeleteTags',
+      resolve: {
+        tag: () => {
+          return tag;
+        }
+      }
+    });
+    modal.result
+      .then((message) => {
+        if (message === 'success') {
+          this.fetchTags();
+        }
+    });
+  }
+
+  showAddModal() {
+    let modal = this.$uibModal.open({
       templateUrl: '/assets/features/tracker/templates/partial/add-preferred-tags-popover.html',
-      size: 'md',
+      size: 'sm',
       backdrop: true,
       keyboard: true,
-      windowTopClass: 'tracker-modal add-preferred-tags-modal'
+      windowTopClass: 'tracker-modal add-preferred-tags-modal',
+      controller: AddPreferredTagsModalCtrl,
+      controllerAs: 'AddTags'
+    });
+    modal.result
+      .then((message) => {
+        if (message === 'success') {
+          this.fetchTags();
+        }
+    });
+  }
+}
+
+function AddPreferredTagsModalCtrl (myTrackerApi, $scope, $state) {
+  'ngInject';
+  this.proceedToNextStep = false;
+
+  function parseTags(tags) {
```

Comment:
We should probably for null or undefined of tags. Not that its a very common use case nevertheless prevents a JS error .